docs/core_ideas.mld

```diff
@@ -0,0 +1,71 @@
+
+  In "normal" OCaml, [List.hd] has type ['a list -> 'a].
+  A call [List.hd []] will throw an exception.
+  By contrast, in the Jane Street libraries, the same function has type ['a list -> 'a option].
+
+  Usually, a function with the suffix [_exn] recovers the original signature, e.g. [List.hd_exn : 'a list -> 'a ].}
+}
+
+If for some reason you {e need} functionality from the OCaml standard library that is not available in Jane Street
+(be sure to triple check), you can use the modules [Stdlib] and [Caml] to access the built-in versions. Currently there
+is only one such usage in the compiler, to use the normal definition of [!=] in the Menhir parser.
+
+There are a few other things we gain from these libraries. The most important idea to understand is {b deriving}.
+
+{2 Deriving functions}
+
+If you look at a type declaration of something like [Ast.typed_expression], you'll notice something curious after the declaration:
+
+{[
+type typed_expression = (typed_expr_meta, fun_kind) expr_with
+[@@deriving sexp, compare, map, hash, fold]
+]}
+
+If you use an editor which supports code completion, you may also notice that the [Ast] module suggests functions
+which are not defined in the actual source text. This is because these functions are created at compile time by
+{{:https://github.com/ocaml-ppx/ppx_deriving}[ppx_deriving]}.The above syntax [[@@deriving ...]] indicates which
+functions we would like to be generated.
+
+These are very helpful - if a type derives [hash], it can automatically be used as keys in Jane Street's hash tables, and
+a type which derives [sexp] can be serialized to Lisp-style S-Expressions. Most of our major types derive at least one function.
+
+{1 "Two Level Types"}
+
+The other curious thing about the types we define in our AST and MIR is they use a trick known as "two level types".
+This allows use to re-use the same type structure with different amounts of metadata attatched (e.g., before typechecking we have
+an [untyped_program] which only has nodes and their source locations, after typechecking we have [typed_program] which features
+nodes, locations, {i and} types, in the same basic tree structure.
+
+The way that this is implemented is a bit non-obvious at first. Essentially,
+many of the tree variant types are parameterized by something that ends up being a placeholder not
+for just metadata but for the recursive type including metadata,
+sometimes called the fixed point.
+So instead of recursively referencing [expression] you would instead reference type parameter ['e],
+which will later be filled in with something like [type expr_with_meta = metadata expression].
+
+
+This takes some getting used to, and also can lead to some unhelpful type signatures in editors such as
+VSCode, due to the fact that abbreviations are not always used. For example, [Expr.Typed.t], the MIR's typed
+expression type, actually has a signature of [Expr.Typed.Meta.t Expr.Fixed.t].
```

Comment:
```suggestion
This takes some getting used to and also can lead to some unhelpful type signatures in editors such as
VSCode, because abbreviations are not always used. For example, [Expr.Typed.t], the MIR's typed
expression type, actually has a signature of [Expr.Typed.Meta.t Expr.Fixed.t].
```